Breast Cancer Classification using Random Forest

In [1]:
from sklearn.datasets import load_breast_cancer
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import roc_auc_score


In [2]:
data=load_breast_cancer()
df=pd.DataFrame(data.data,columns=data.feature_names)
df['target']=data.target
print(df.head())

   mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0             

In [5]:
print("Dataset Shape:")
print(df.shape)
print("\nTarget Distribution:")
df["target"].value_counts()


(569, 31)


In [21]:
X = df.drop("target", axis=1)
y = df["target"]
print(X.shape)
print(y.shape)

(569, 30)
(569,)


In [7]:

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(X_train.shape)
print(X_test.shape)

print(y_train.shape)
print(y_test.shape)

(455, 30)
(114, 30)
(455,)
(114,)


In [18]:
rf = RandomForestClassifier(random_state=42)
rf.fit(X_train, y_train)

y_pred_rf = rf.predict(X_test)
print(RandomForestClassifier)
print(y_pred_rf[:10])
AUC_rf=accuracy_score(y_test,y_pred_rf)
print("AUC_rf=",AUC_rf)
y_prob_rf = rf.predict_proba(X_test)[:, 1]
roc_auc_rf =roc_auc_score(y_test, y_prob_rf)
print("roc_auc_rf =", roc_auc_rf)

<class 'sklearn.ensemble._forest.RandomForestClassifier'>
[1 0 0 1 1 0 0 0 0 1]
AUC_rf= 0.9649122807017544
roc_auc_rf = 0.9952505732066819


In [19]:

xgb=XGBClassifier(random_state=42)
xgb.fit(X_train,y_train)

y_pred_xgb=xgb.predict(X_test)
print(XGBClassifier)
print(y_pred_xgb[:10])
AUC_xgb=accuracy_score(y_test,y_pred_xgb)
print("AUC_xgb=",AUC_xgb)
y_prob_xgb = xgb.predict_proba(X_test)[:,1]
roc_auc_xgb = roc_auc_score(y_test,y_prob_xgb)
print("roc_auc_xgb =",auc_xgb)

<class 'xgboost.sklearn.XGBClassifier'>
[1 0 0 1 1 0 0 0 0 1]
AUC_xgb= 0.956140350877193
roc_auc_xgb = 0.9908286930887652


In [20]:
table=pd.DataFrame({
    "Model":["Random Forest","XGBoost"],
    "AUC":[AUC_rf,AUC_xgb],
    "ROC_AUC":[roc_auc_rf,roc_auc_xgb]
})
print(table)

           Model       AUC   ROC_AUC
0  Random Forest  0.964912  0.995251
1        XGBoost  0.956140  0.990829


In [ ]:
importance_df = pd.DataFrame({
    "Feature": X.columns,
    "Importance": rf.feature_importances_
})

importance_df.sort_values(
    by="Importance",
    ascending=False
).head(10)


,Feature,Importance
23,worst area,0.153892
27,worst concave points,0.144663
7,mean concave points,0.106210
20,worst radius,0.077987
6,mean concavity,0.068001
22,worst perimeter,0.067115
2,mean perimeter,0.053270
0,mean radius,0.048703
3,mean area,0.047555
26,worst concavity,0.031802


# Conclusion

Random Forest achieved:

- Accuracy = 96.5%
- AUC = 0.995

XGBoost achieved:

- Accuracy = 95.6%
- AUC = 0.991

Both models demonstrated excellent performance.

Random Forest slightly outperformed XGBoost on this dataset.